feat: container page table component

deboer-tim opened the pull request:
### What does this PR do?

Switch the Containers page to use the Table component. This is a final piece
of technical debt adopting the Table component, and gives a few advantages:
- Table expands/collapses to fit the display width (used to extend off the
  right edge with a scrollbar).
- Sorting of name column to start (wanted to focus on core aspects and get
  this in before sorting other columns).
- Consistent table UI with other pages.
- Basic support for light mode.

The containers table included ContainerGroupInfoUI parents of three types
(pods, compose groups, and standalone containers), all with ContainerInfoUI
children. To avoid UI issues the parent ContainerGroupInfoUI was skipped
when drawing standalone containers (i.e. it only draws the child container
object in this one case).

Since the Table component doesn't support this unique behaviour, the least
invasive/simplest option was to create a derived list where standalone
containers are mapped to actual containers.

Found an issue where Svelte was updating constantly; turns out this was
an existing bug in the timeout when no containers are running.

Ignore changes to Table - reviewed separately under [#7587](https://github.com/deboer-tim/desktop/issues/7587).

### Screenshot / video of UI

Before:

<img width="1162" alt="Screenshot 2024-06-11 at 11 29 19 AM" src="https://github.com/containers/podman-desktop/assets/19958075/391371b5-d558-4687-aa95-1d5ea4905b78">

After:

<img width="1162" alt="Screenshot 2024-06-11 at 11 49 38 AM" src="https://github.com/containers/podman-desktop/assets/19958075/3915c707-9b74-4d43-9bf8-000dde6a2a2d">

### What issues does this PR fix or reference?

Fixes #4843.

### How to test this PR?

Try Containers page for regular containers, pods, and compose.

- [x] Tests are covering the bug fix or the new feature

cdrage commented on packages/renderer/src/lib/container/ContainerList.svelte:
```diff
@@ -463,204 +485,15 @@ function setStoppedFilter() {
+      data="{containerGroups}"
+      columns="{columns}"
+      row="{row}"
+      on:update="{() => (containerGroups = [...containerGroups])}">
```
should we add `defaultSortColumn="Name"` ?

deboer-tim replied:
Done in latest update.

deboer-tim commented:
Rebased and updated, ready for full review!

deboer-tim commented:
Rebased on latest #7587 to avoid build failures.

cdrage commented:
Noticed some things feature wise before I dive deeper into the code:


**Light mode not working:**
![Screenshot 2024-06-11 at 2 48 54 PM](https://github.com/containers/podman-desktop/assets/6422176/80ba48e3-74e4-4b80-81cb-48a43d57e987)

**Compose isn't working, but pods are:**

https://github.com/containers/podman-desktop/assets/6422176/1ef95794-fd31-42b3-828d-61cfb09b0d84

To replicate:

```sh
git clone https://github.com/cdrage/composefiles
cd guestbook-go
podman compose up -d
```

And then try it in the interface, the containers do not open or close.

**Missing operators on Status, Environment, Image and Age:**
![Screenshot 2024-06-11 at 2 50 57 PM](https://github.com/containers/podman-desktop/assets/6422176/981a2ed8-b4b8-435d-a43f-ba2126f85741)

axel7083 commented on packages/renderer/src/lib/container/ContainerColumnActions.svelte:
```diff
@@ -0,0 +1,52 @@
+export let object: ContainerInfoUI | ContainerGroupInfoUI;
+</script>
+
+{#if object.type === ContainerGroupInfoTypeUI.POD && object.engineId && object.id && object.shortId && object.status && object.engineName && object.humanCreationDate && object.created}
```
I can see that we often repeat the check `type === ContainerGroupInfoTypeUI.POD`, might be interesting to have an utility method

isPod(object: unknown): object is ContainerGroupInfoUI or something similar ? (not in this PR, just a general feeling)

deboer-tim replied:
Yeah, I think we do this (and compose) several places in the code so might be nice to do this.

axel7083 commented on packages/renderer/src/lib/container/ContainerColumnStatus.svelte:
```diff
@@ -0,0 +1,16 @@
+export let object: ContainerInfoUI | ContainerGroupInfoUI;
+</script>
+
+{#if 'type' in object && object.type === ContainerGroupInfoTypeUI.POD}
+  <StatusIcon icon="{PodIcon}" status="{object.status}" />
+{:else if 'type' in object && object.type === ContainerGroupInfoTypeUI.COMPOSE}
+  <StatusIcon icon="{PodIcon}" status="{object.status}" />
+{:else if 'state' in object}
+  <StatusIcon icon="{object.icon}" status="{object.state}" />
+{/if}
```
```suggestion
{#if 'type' in object && typeof object.type === 'string'}
  <StatusIcon icon="{PodIcon}" status="{object.status}" />
{:else if 'state' in object && typeof object.type === 'string'}
  <StatusIcon icon="{object.icon}" status="{object.state}" />
{/if}
```

line 11 and 13 are the same, might be nice to simplify this

deboer-tim replied:
I simplified in new commit, but just to `#if 'type' in object`. I assume you meant `typeof object.state` in the second case?  There are only two types here (ContainerInfoUI | ContainerGroupInfoUI) and type/state are each only on one, so I didn't really think checking the type of each property was necessary.

axel7083 replied:
Yeah having it everywhere would be redundant

deboer-tim replied:
Added isContainerGroupInfoUI/isContainerInfoUI helper utils, and they use this logic.

axel7083 commented on packages/renderer/src/lib/container/ContainerColumnName.svelte:
```diff
@@ -0,0 +1,63 @@
+}
+</script>
+
+{#if 'type' in object && (object.type === ContainerGroupInfoTypeUI.POD || object.type === ContainerGroupInfoTypeUI.COMPOSE)}
```
`'type' in object && (object.type === ContainerGroupInfoTypeUI.POD || object.type === ContainerGroupInfoTypeUI.COMPOSE)` could be replaced by an utility method isContainerGroup() or something similar

deboer-tim replied:
I forgot that I simplified and avoided this ever being called with a standalone container parent, so I was able to remove `object.type ===` in the new commit. But yes, a utility method might be useful in future.

axel7083 commented on packages/renderer/src/lib/container/ContainerColumnName.svelte:
```diff
@@ -0,0 +1,63 @@
+    return;
+  }
+  if (containerGroup.type === ContainerGroupInfoTypeUI.POD) {
+    router.goto(`/pods/podman/${encodeURI(containerGroup.name)}/${encodeURIComponent(containerGroup.engineId)}/logs`);
```
I am wondering if we should not uses the `handleNavigation` from 

https://github.com/containers/podman-desktop/blob/54cdb5cd43cae0162f533221a1f4e0bd74dce66a/packages/renderer/src/navigation.ts#L32

to avoid hardcoding the urls in the code in multiple places

> Not a blocker

deboer-tim replied:
Yes, we should do this. I opened #4966 a while ago, I think we should target a separate PR (or PRs) to clean this up in other places as well.

axel7083 commented on packages/renderer/src/lib/container/ContainerColumnName.svelte:
```diff
@@ -0,0 +1,63 @@
+  return result;
+}
+
+function openGroupDetails(containerGroup: ContainerGroupInfoUI) {
```
Why is the linter not complaining about the return type not provided ?

deboer-tim replied:
Good catch, I didn't notice while doing copy/paste. Maybe we don't check .svelte in renderer? I fixed all that I could find in the new commit.

axel7083 commented on packages/renderer/src/lib/container/ContainerColumnImage.svelte:
```diff
@@ -0,0 +1,11 @@
+export let object: ContainerInfoUI | ContainerGroupInfoUI;
+</script>
+
+{#if 'image' in object}
```
```suggestion
{#if 'image' in object && typeof object.image === 'string'}
```

deboer-tim replied:
Ditto to previous comment, there are only two types and image only exists in one of them, so what's the added benefit of checking the property type?

axel7083 replied:
type safety, never trust pattern

deboer-tim replied:
As per other comment, I've added isContainerGroupInfoUI/isContainerInfoUI helper utils which do this.

axel7083 commented on packages/renderer/src/lib/container/ContainerColumnAge.svelte:
```diff
@@ -0,0 +1,12 @@
+export let object: ContainerInfoUI | ContainerGroupInfoUI;
+</script>
+
+{#if 'state' in object}
```
```suggestion
{#if 'state' in object && typeof object.state === 'string'} 
```

axel7083 replied:
I am wondering if typescript should not complain here as state would be unknown type before checking its type

deboer-tim replied:
Only ContainerInfoUI has state and it's a string there though?

axel7083 replied:
what if I am sending you a { state: true }

deboer-tim replied:
As per other comment, I've added isContainerGroupInfoUI/isContainerInfoUI helper utils which do this.

axel7083 reviewed (approved):
Amazing job, tested locally and loved it! 😄 🚀

deboer-tim commented:
Rebased to pick up Table fix and other changes in main. Will respond to the rest of the comments inline, but ready for another review.

> Noticed some things feature wise before I dive deeper into the code:
>
> Light mode not working:

I started this well before we were fixing columns :( and was going to do that next. Simple enough though, added in new commit.

> **Compose isn't working, but pods are:**

Rebased on fix from #7587.

> **Missing operators on Status, Environment, Image and Age:**

I assume you meant the sorting? This was intentional, I was trying to keep this PR focused on the main port and small(er) to get it in, then we can add new capability.

deboer-tim commented on packages/renderer/src/lib/container/ContainerColumnAge.svelte:
```diff
@@ -0,0 +1,12 @@
+<script lang="ts">
```
It'd still need a wrapper b/c this column shows text like 'Starting' too, but otherwise yes. There is no date exposed in ContainerInfoUI though so I didn't want to include here (and I think pods have the same issue).

axel7083 replied:
you could use rendermapping to convert it to a date object

deboer-tim replied:
Not directly, since the column also uses the status. (I agree with direction, was saying more that it should be a future PR)

benoitf reviewed (commented):
we're missing functions returning `is ContainerInfoUI` and `is ContainerGroupInfoUI`
as you're checking for properties like `type` or `state` in svelte objects while we shouldn't.

```
it should be const containerInfo = isContainerInfo(object)
if (containerInfo) {
  // here you can use containerInfo.state as there is the cast
}
```

and probably have renderer dedicated to containers and groups

so the final renderer should only have object being `ContainerInfoUI` or `ContainerGroupInfoUI` not the union type

axel7083 commented on packages/renderer/src/lib/container/ContainerColumnActions.svelte:
```diff
@@ -0,0 +1,24 @@
+{#if containerUtils.isContainerGroupInfoUI(object)}
+  {#if object.type === ContainerGroupInfoTypeUI.POD}
+    <ContainerColumnActionsPod object="{object}" />
+  {/if}
+  {#if object.type === ContainerGroupInfoTypeUI.COMPOSE}
```
```suggestion
  {:else if object.type === ContainerGroupInfoTypeUI.COMPOSE}
```

An object type cannot be POD and COMPOSE at the same time

deboer-tim replied:
Fixed.

axel7083 commented on packages/renderer/src/lib/container/ContainerColumnActionsContainer.svelte:
```diff
@@ -0,0 +1,13 @@
+{#if object.actionError}
+  <ErrorMessage error="{object.actionError}" icon />
+{/if}
+<ContainerActions container="{object}" dropdownMenu="{true}" on:update />
```
```suggestion
<ContainerActions container="{object}" dropdownMenu="{true}" />
```

deboer-tim replied:
Fixed.

axel7083 commented on packages/renderer/src/lib/container/ContainerColumnAge.svelte:
```diff
@@ -0,0 +1,15 @@
+
+export let object: ContainerInfoUI | ContainerGroupInfoUI;
+
+const containerUtils = new ContainerUtils();
```
I would argue that we could have static method instead of having a dedicated class for this, I know @benoitf have a preference for it, as it _ease_ the testing _in some cases_ but here I feel like it adds complexity when reading it, as we might think ContainerUtils has internal logic, as it is a class

deboer-tim replied:
Yeah, I added here b/c well, everything else related is here. Could do static functions in the same file? Easy switch if someone prefers.

axel7083 replied:
yeah I would rather have static method where ContainerUtils is defined ?

axel7083 replied:
I just review the `containers-utils` file and it should probably be kept like you did for consistency, so forget my comment

axel7083 commented on packages/renderer/src/lib/container/ContainerList.svelte:
```diff
@@ -1,11 +1,10 @@
-import { faChevronDown, faChevronRight, faPlusCircle, faTrash } from '@fortawesome/free-solid-svg-icons';
-import { Button, Checkbox, ErrorMessage, FilteredEmptyScreen, Modal, NavPage } from '@podman-desktop/ui-svelte';
+import { faPlusCircle, faTrash } from '@fortawesome/free-solid-svg-icons';
+import { Button, FilteredEmptyScreen, Modal, NavPage, Table, TableColumn, TableRow } from '@podman-desktop/ui-svelte';
```
<3 so many UI components

deboer-tim replied:
Yup :(

deboer-tim commented:
> we're missing functions returning `is ContainerInfoUI` and `is ContainerGroupInfoUI` as you're checking for properties like `type` or `state` in svelte objects while we shouldn't.
> 
> ```
> it should be const containerInfo = isContainerInfo(object)
> if (containerInfo) {
>   // here you can use containerInfo.state as there is the cast
> }
> ```

Ok, added these in new commit to avoid doing this in Svelte. There are still a couple cases where .type is checked for pod vs compose and we could do util for these too, but let me know what you think.

> and probably have renderer dedicated to containers and groups
> 
> so the final renderer should only have object being `ContainerInfoUI` or `ContainerGroupInfoUI` not the union type

I struggled with this because in the 'worst case' it means four files for every column: initial switcher column, plus components for pod, compose, and container. I did this for actions since there is different logic for each one, but not for age because there's just one if and div. Again, let me know if you think it should go further.

benoitf reviewed (approved):
works for me now
